// @mantine/charts/cjs/Sparkline/Sparkline.cjs
'use client';
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var react = require('react');
var recharts = require('recharts');
var core = require('@mantine/core');
require('../AreaChart/AreaChart.cjs');
var AreaGradient = require('../AreaChart/AreaGradient.cjs');

const classes = {};
const defaultProps = {
  withGradient: true,
  connectNulls: true,
  fillOpacity: 0.6,
  strokeWidth: 2,
  curveType: "linear"
};
function getTrendColor(data, trendColors) {
  const first = data[0];
  const last = data[data.length - 1];
  if (first === null || last === null) {
    return trendColors.neutral || trendColors.positive;
  }
  if (first < last) {
    return trendColors.positive;
  }
  if (first > last) {
    return trendColors.negative;
  }
  return trendColors.neutral || trendColors.positive;
}
const varsResolver = core.createVarsResolver(
  (theme, { color, data, trendColors }) => ({
    root: {
      "--chart-color": trendColors ? core.getThemeColor(getTrendColor(data, trendColors), theme) : color ? core.getThemeColor(color, theme) : void 0
    }
  })
);
const Sparkline = core.factory((_props, ref) => {
  const props = core.useProps("Sparkline", defaultProps, _props);
  const {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    data,
    withGradient,
    fillOpacity,
    curveType,
    strokeWidth,
    trendColors,
    connectNulls,
    areaProps,
    ...others
  } = props;
  const getStyles = core.useStyles({
    name: "Sparkline",
    classes,
    props,
    className,
    style,
    classNames,
    styles,
    unstyled,
    vars,
    varsResolver
  });
  const id = react.useId();
  const mappedData = react.useMemo(() => data.map((value, index) => ({ value, index })), [data]);
  return /* @__PURE__ */ jsxRuntime.jsx(core.Box, { ref, ...getStyles("root"), ...others, dir: "ltr", children: /* @__PURE__ */ jsxRuntime.jsx(recharts.ResponsiveContainer, { children: /* @__PURE__ */ jsxRuntime.jsxs(recharts.AreaChart, { data: mappedData, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      recharts.Area,
      {
        dataKey: "value",
        type: curveType,
        fill: `url(#${id})`,
        stroke: "var(--chart-color, var(--mantine-color-blue-filled))",
        isAnimationActive: false,
        connectNulls,
        strokeWidth,
        fillOpacity: 1,
        ...areaProps
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx("defs", { children: /* @__PURE__ */ jsxRuntime.jsx(
      AreaGradient.AreaGradient,
      {
        id,
        color: "var(--chart-color, var(--mantine-color-blue-filled))",
        fillOpacity,
        withGradient
      }
    ) })
  ] }) }) });
});
Sparkline.displayName = "@mantine/charts/Sparkline";
Sparkline.classes = classes;

exports.Sparkline = Sparkline;
