// @mantine/charts/cjs/ChartTooltip/ChartTooltip.cjs
'use client';
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var core = require('@mantine/core');
var getSeriesLabels = require('../utils/get-series-labels/get-series-labels.cjs');
var ChartTooltip_module = require('./ChartTooltip.module.css.cjs');

function updateChartTooltipPayload(payload) {
  return payload.map((item) => {
    const newDataKey = item.name.split(".").pop();
    return {
      ...item,
      name: newDataKey
    };
  });
}
function getFilteredChartTooltipPayload(payload, segmentId) {
  const duplicatesFilter = updateChartTooltipPayload(
    payload.filter((item) => item.fill !== "none" || !item.color)
  );
  if (!segmentId) {
    return duplicatesFilter;
  }
  return duplicatesFilter.filter((item) => item.name === segmentId);
}
function getData(item, type) {
  if (type === "radial" || type === "scatter") {
    if (Array.isArray(item.value)) {
      return item.value[1] - item.value[0];
    }
    return item.value;
  }
  if (Array.isArray(item.payload[item.dataKey])) {
    return item.payload[item.dataKey][1] - item.payload[item.dataKey][0];
  }
  return item.payload[item.dataKey];
}
const defaultProps = {
  type: "area",
  showColor: true
};
const ChartTooltip = core.factory((_props, ref) => {
  const props = core.useProps("ChartTooltip", defaultProps, _props);
  const {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    payload,
    label,
    unit,
    type,
    segmentId,
    mod,
    series,
    valueFormatter,
    showColor,
    ...others
  } = props;
  const theme = core.useMantineTheme();
  const getStyles = core.useStyles({
    name: "ChartTooltip",
    classes: ChartTooltip_module,
    props,
    className,
    style,
    classNames,
    styles,
    unstyled
  });
  if (!payload) {
    return null;
  }
  const filteredPayload = getFilteredChartTooltipPayload(payload, segmentId);
  const scatterLabel = type === "scatter" ? payload[0]?.payload?.name : null;
  const labels = getSeriesLabels.getSeriesLabels(series);
  const _label = label || scatterLabel;
  const items = filteredPayload.map((item) => /* @__PURE__ */ jsxRuntime.jsxs("div", { "data-type": type, ...getStyles("tooltipItem"), children: [
    /* @__PURE__ */ jsxRuntime.jsxs("div", { ...getStyles("tooltipItemBody"), children: [
      showColor && /* @__PURE__ */ jsxRuntime.jsx(
        core.ColorSwatch,
        {
          color: core.getThemeColor(item.color, theme),
          size: 12,
          ...getStyles("tooltipItemColor"),
          withShadow: false
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx("div", { ...getStyles("tooltipItemName"), children: labels[item.name] || item.name })
    ] }),
    /* @__PURE__ */ jsxRuntime.jsxs("div", { ...getStyles("tooltipItemData"), children: [
      typeof valueFormatter === "function" ? valueFormatter(getData(item, type)) : getData(item, type),
      unit || item.unit
    ] })
  ] }, item?.key ?? item.name));
  return /* @__PURE__ */ jsxRuntime.jsxs(core.Box, { ...getStyles("tooltip"), mod: [{ type }, mod], ref, ...others, children: [
    _label && /* @__PURE__ */ jsxRuntime.jsx("div", { ...getStyles("tooltipLabel"), children: _label }),
    /* @__PURE__ */ jsxRuntime.jsx("div", { ...getStyles("tooltipBody"), children: items })
  ] });
});
ChartTooltip.displayName = "@mantine/charts/ChartTooltip";

exports.ChartTooltip = ChartTooltip;
exports.getFilteredChartTooltipPayload = getFilteredChartTooltipPayload;
