// @mantine/charts/cjs/BubbleChart/BubbleChart.cjs
'use client';
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var recharts = require('recharts');
var core = require('@mantine/core');
var gridChart_module = require('../grid-chart.module.css.cjs');

function getDomain(data, key) {
  const values = data.map((item) => item[key]);
  return [Math.min(...values), Math.max(...values)];
}
function BubbleChartTooltip({
  active,
  payload,
  getStyles,
  dataKey,
  valueFormatter
}) {
  if (active && payload && payload.length) {
    const data = payload[0] && payload[0].payload;
    return /* @__PURE__ */ jsxRuntime.jsx("div", { ...getStyles("tooltip"), children: /* @__PURE__ */ jsxRuntime.jsxs(core.Group, { justify: "space-between", children: [
      /* @__PURE__ */ jsxRuntime.jsx(core.Text, { fz: "sm", children: data[dataKey.x] }),
      /* @__PURE__ */ jsxRuntime.jsx(core.Text, { fz: "sm", children: valueFormatter ? valueFormatter(data[dataKey.z]) : data[dataKey.z] })
    ] }) });
  }
  return null;
}
const defaultProps = {
  color: "blue.6",
  withTooltip: true
};
const varsResolver = core.createVarsResolver((theme, { textColor, gridColor }) => ({
  root: {
    "--chart-text-color": textColor ? core.getThemeColor(textColor, theme) : void 0,
    "--chart-grid-color": gridColor ? core.getThemeColor(gridColor, theme) : void 0
  }
}));
const BubbleChart = core.factory((_props, ref) => {
  const props = core.useProps("BubbleChart", defaultProps, _props);
  const {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    data,
    xAxisProps,
    yAxisProps,
    zAxisProps,
    tooltipProps,
    scatterProps,
    color,
    label,
    withTooltip,
    dataKey,
    range,
    valueFormatter,
    ...others
  } = props;
  const theme = core.useMantineTheme();
  const getStyles = core.useStyles({
    name: "BubbleChart",
    classes: gridChart_module,
    props,
    className,
    style,
    classNames,
    styles,
    unstyled,
    vars,
    varsResolver
  });
  return /* @__PURE__ */ jsxRuntime.jsx(core.Box, { ref, ...getStyles("root"), ...others, children: /* @__PURE__ */ jsxRuntime.jsx(recharts.ResponsiveContainer, { children: /* @__PURE__ */ jsxRuntime.jsxs(recharts.ScatterChart, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      recharts.XAxis,
      {
        type: "category",
        dataKey: dataKey.x,
        interval: 0,
        tick: { fontSize: 12, fill: "currentColor" },
        tickLine: { transform: "translate(0, -6)", stroke: "currentColor" },
        stroke: "currentColor",
        ...getStyles("axis"),
        ...xAxisProps
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      recharts.YAxis,
      {
        type: "number",
        dataKey: dataKey.y,
        height: 10,
        width: label ? void 0 : 0,
        tick: false,
        tickLine: false,
        axisLine: false,
        label: { value: label, position: "insideRight", fontSize: 12, fill: "currentColor" },
        ...getStyles("axis"),
        ...yAxisProps
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      recharts.ZAxis,
      {
        type: "number",
        dataKey: dataKey.z,
        domain: getDomain(data, dataKey.z),
        range,
        ...zAxisProps
      }
    ),
    withTooltip && /* @__PURE__ */ jsxRuntime.jsx(
      recharts.Tooltip,
      {
        animationDuration: 100,
        isAnimationActive: false,
        cursor: { stroke: "var(--chart-grid-color)", strokeWidth: 1, strokeDasharray: "3 3" },
        content: (payload) => /* @__PURE__ */ jsxRuntime.jsx(
          BubbleChartTooltip,
          {
            dataKey,
            active: payload.active,
            payload: payload.payload,
            getStyles,
            valueFormatter
          }
        ),
        ...tooltipProps
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(
      recharts.Scatter,
      {
        data,
        fill: core.getThemeColor(color, theme),
        isAnimationActive: false,
        ...scatterProps
      }
    )
  ] }) }) });
});
BubbleChart.displayName = "@mantine/charts/BubbleChart";
BubbleChart.classes = gridChart_module;

exports.BubbleChart = BubbleChart;
