// @mantine/spotlight/cjs/index.cjs
'use strict';

var spotlight_store = require('./spotlight.store.cjs');
var isActionsGroup = require('./is-actions-group.cjs');
var Spotlight = require('./Spotlight.cjs');
var SpotlightAction = require('./SpotlightAction.cjs');
var SpotlightActionsGroup = require('./SpotlightActionsGroup.cjs');
var SpotlightActionsList = require('./SpotlightActionsList.cjs');
var SpotlightEmpty = require('./SpotlightEmpty.cjs');
var SpotlightFooter = require('./SpotlightFooter.cjs');
var SpotlightSearch = require('./SpotlightSearch.cjs');



exports.closeSpotlight = spotlight_store.closeSpotlight;
exports.createSpotlight = spotlight_store.createSpotlight;
exports.createSpotlightStore = spotlight_store.createSpotlightStore;
exports.openSpotlight = spotlight_store.openSpotlight;
exports.spotlight = spotlight_store.spotlight;
exports.toggleSpotlight = spotlight_store.toggleSpotlight;
exports.useSpotlight = spotlight_store.useSpotlight;
exports.isActionsGroup = isActionsGroup.isActionsGroup;
exports.Spotlight = Spotlight.Spotlight;
exports.SpotlightAction = SpotlightAction.SpotlightAction;
exports.SpotlightActionsGroup = SpotlightActionsGroup.SpotlightActionsGroup;
exports.SpotlightActionsList = SpotlightActionsList.SpotlightActionsList;
exports.SpotlightEmpty = SpotlightEmpty.SpotlightEmpty;
exports.SpotlightFooter = SpotlightFooter.SpotlightFooter;
exports.SpotlightSearch = SpotlightSearch.SpotlightSearch;
