// @mantine/charts/cjs/ChartLegend/ChartLegend.cjs
'use client';
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var core = require('@mantine/core');
var getSeriesLabels = require('../utils/get-series-labels/get-series-labels.cjs');
var ChartLegend_module = require('./ChartLegend.module.css.cjs');

function updateChartLegendPayload(payload) {
  return payload.map((item) => {
    const newDataKey = item.dataKey.split(".").pop();
    return {
      ...item,
      dataKey: newDataKey,
      payload: {
        ...item.payload,
        name: newDataKey,
        dataKey: newDataKey
      }
    };
  });
}
function getFilteredChartLegendPayload(payload) {
  return updateChartLegendPayload(payload.filter((item) => item.color !== "none"));
}
const defaultProps = {};
const ChartLegend = core.factory((_props, ref) => {
  const props = core.useProps("ChartLegend", defaultProps, _props);
  const {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    payload,
    onHighlight,
    legendPosition,
    mod,
    series,
    showColor,
    centered,
    ...others
  } = props;
  const getStyles = core.useStyles({
    name: "ChartLegend",
    classes: ChartLegend_module,
    props,
    className,
    style,
    classNames,
    styles,
    unstyled
  });
  if (!payload) {
    return null;
  }
  const filteredPayload = getFilteredChartLegendPayload(payload);
  const labels = getSeriesLabels.getSeriesLabels(series);
  const items = filteredPayload.map((item, index) => /* @__PURE__ */ jsxRuntime.jsxs(
    "div",
    {
      ...getStyles("legendItem"),
      onMouseEnter: () => onHighlight(item.dataKey),
      onMouseLeave: () => onHighlight(null),
      "data-without-color": showColor === false || void 0,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          core.ColorSwatch,
          {
            color: item.color,
            size: 12,
            ...getStyles("legendItemColor"),
            withShadow: false
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx("p", { ...getStyles("legendItemName"), children: labels[item.dataKey] || item.dataKey })
      ]
    },
    index
  ));
  return /* @__PURE__ */ jsxRuntime.jsx(
    core.Box,
    {
      ref,
      mod: [{ position: legendPosition, centered }, mod],
      ...getStyles("legend"),
      ...others,
      children: items
    }
  );
});
ChartLegend.displayName = "@mantine/charts/ChartLegend";

exports.ChartLegend = ChartLegend;
exports.getFilteredChartLegendPayload = getFilteredChartLegendPayload;
