// @mantine/tiptap/cjs/index.cjs
'use strict';

var Link = require('./extensions/Link.cjs');
var TaskList = require('./extensions/TaskList.cjs');
var RichTextEditor = require('./RichTextEditor.cjs');
var RichTextEditor_context = require('./RichTextEditor.context.cjs');
var labels = require('./labels.cjs');
var controls = require('./RichTextEditorControl/controls.cjs');
var RichTextEditorLinkControl = require('./RichTextEditorControl/RichTextEditorLinkControl.cjs');
var RichTextEditorColorPickerControl = require('./RichTextEditorControl/RichTextEditorColorPickerControl.cjs');
var RichTextEditorColorControl = require('./RichTextEditorControl/RichTextEditorColorControl.cjs');
var RichTextEditorControlsGroup = require('./RichTextEditorControlsGroup/RichTextEditorControlsGroup.cjs');
var RichTextEditorControl = require('./RichTextEditorControl/RichTextEditorControl.cjs');
var RichTextEditorContent = require('./RichTextEditorContent/RichTextEditorContent.cjs');



exports.Link = Link.Link;
exports.getTaskListExtension = TaskList.getTaskListExtension;
exports.RichTextEditor = RichTextEditor.RichTextEditor;
exports.useRichTextEditorContext = RichTextEditor_context.useRichTextEditorContext;
exports.DEFAULT_LABELS = labels.DEFAULT_LABELS;
exports.AlignCenterControl = controls.AlignCenterControl;
exports.AlignJustifyControl = controls.AlignJustifyControl;
exports.AlignLeftControl = controls.AlignLeftControl;
exports.AlignRightControl = controls.AlignRightControl;
exports.BlockquoteControl = controls.BlockquoteControl;
exports.BoldControl = controls.BoldControl;
exports.BulletListControl = controls.BulletListControl;
exports.ClearFormattingControl = controls.ClearFormattingControl;
exports.CodeBlockControl = controls.CodeBlockControl;
exports.CodeControl = controls.CodeControl;
exports.H1Control = controls.H1Control;
exports.H2Control = controls.H2Control;
exports.H3Control = controls.H3Control;
exports.H4Control = controls.H4Control;
exports.H5Control = controls.H5Control;
exports.H6Control = controls.H6Control;
exports.HighlightControl = controls.HighlightControl;
exports.HrControl = controls.HrControl;
exports.ItalicControl = controls.ItalicControl;
exports.OrderedListControl = controls.OrderedListControl;
exports.RedoControl = controls.RedoControl;
exports.StrikeThroughControl = controls.StrikeThroughControl;
exports.SubscriptControl = controls.SubscriptControl;
exports.SuperscriptControl = controls.SuperscriptControl;
exports.TaskListControl = controls.TaskListControl;
exports.TaskListLiftControl = controls.TaskListLiftControl;
exports.TaskListSinkControl = controls.TaskListSinkControl;
exports.UnderlineControl = controls.UnderlineControl;
exports.UndoControl = controls.UndoControl;
exports.UnlinkControl = controls.UnlinkControl;
exports.UnsetColorControl = controls.UnsetColorControl;
exports.RichTextEditorLinkControl = RichTextEditorLinkControl.RichTextEditorLinkControl;
exports.RichTextEditorColorPickerControl = RichTextEditorColorPickerControl.RichTextEditorColorPickerControl;
exports.RichTextEditorColorControl = RichTextEditorColorControl.RichTextEditorColorControl;
exports.RichTextEditorControlsGroup = RichTextEditorControlsGroup.RichTextEditorControlsGroup;
exports.RichTextEditorControl = RichTextEditorControl.RichTextEditorControl;
exports.RichTextEditorContent = RichTextEditorContent.RichTextEditorContent;
